### DISCUSSION TITLE
How to have default sorting for table?

### DISCUSSION THREAD

**Discussion by minthemiddle** [filamentphp/filament#261] [Help]
Say I have a blog post resource with `created_at`.
How do I make sure that in the table view for the resource (`/admin/resources/posts`) I sort by default with `created_at > desc`?

`sortable()` does not solve it as you have to manually click on it.

Edit: Just a related PR to this: https://github.com/laravel-filament/filament/pull/231

**Comment by csikoszoltan** [comment]
Hello,
have you found a solution to this question?

**Comment by YassinoB** [comment]
**Can you check this code :** 
return $table->columns([
                Columns\Text::make('created_at')
                    ->primary()
                    ->date()
                    ->sortable(),
                Columns\Text::make('updated_at')
                    ->primary()
                    ->date()
                    ->sortable(),
            ])->defaultSort('created_at', 'desc')

@csikoszoltan

**Reply by csikoszoltan** [reply]
It works, thanks! :)

**Reply by TuenTuenna** [reply]
really thanks!

**Reply by aksalsf** [reply]
How to apply it with Livewire getTableColumns?

**Reply by danharrin** [reply]
`getDefaultTableSortColumn()` and `getDefaultTableSortDirection()` for the table builder

**Reply by aksalsf** [reply]
Worked. Thanks!

**Reply by MechJosh0** [reply]
Thanks, @danharrin. Just so you're aware this isn't available in the documentation (or I'm being blind), it'd be very helpful if it was for others 👍

**Reply by danharrin** [reply]
I think its only in the admin panel docs atm, needs to be added to the table builder docs too if you want to PR https://filamentphp.com/docs/2.x/admin/resources/listing-records#sorting-a-column-by-default

**Reply by MechJosh0** [reply]
@danharrin Opened a PR for it https://github.com/filamentphp/filament/pull/3229

**Reply by tahinuralam** [reply]
> ->defaultSort('created_at', 'desc')

how to create table footer row

**Reply by Junior-Shyko** [reply]
Deu certinho aqui para o meu projeto

**Comment by judgej** [comment]
So adding to this really old discussion, the non-panel defaults are documented here: https://filamentphp.com/docs/2.x/tables/columns/getting-started#sorting

Using it with filament 2, I notice the default ordering works to order the table, but the table heading does not reflect the ordering with the downward chevron (⌄) remaining greyed out to sugnify "order not chosen".

I've not raised this as an issue, but can if it is a real issue and not just my app.

**Reply by danharrin** [reply]
This is deliberate behaviour after lots of user feedback and testing. The default sort is separate to triggering a sort on a single column manually. The default sort doesn't even need to be associated with a column that exists on the table, and it can apply to multiple columns.

**Reply by bhattji** [reply]
How to apply on multiple columns? some ascending some descending?

**Reply by gludovatza** [reply]
@bhattji 
`->defaultSort(fn ($query) => $query->orderBy('created_at', 'asc')->orderBy('last_login_at', 'desc'))`
via: https://github.com/filamentphp/filament/discussions/2626#discussioncomment-4125485

**Reply by bhattji** [reply]
Can I add a foreign-key in the parent table? for example if I have a table called cities, which has  a state_id, in states I have a country_id can I add country_id in sort?

**Reply by gludovatza** [reply]
@bhattji 
I think yes, but try it!

**Reply by bhattji** [reply]
I tried it, doesn't work directly, needs help

**Reply by SanderMuller** [reply]
@danharrin I DO want to show it visually in the UI since I am sorting by a column. Our QA flagged the table as a bug since you need to click twice for it to start sorting, which is very odd, and I agree.

Although I understand that you may not want to highlight this by default, it would be nice to have an option to enable it. I tried to modify a lot of different things both in the Filament code and my own code but cannot find a way to make it work the way I want. Also can't find a solution online.

**Reply by danharrin** [reply]
Try this in a service provider

```php
FilamentIcon::register(['tables::header-cell.sort-button' => 'heroicon-m-chevron-up-down'])
```

**Reply by SanderMuller** [reply]
I think I got it working via

```php
    use InteractsWithForms,
        InteractsWithTable {
        bootedInteractsWithTable as baseBootedInteractsWithTable;
    }

    public function bootedInteractsWithTable(): void
    {
        $this->baseBootedInteractsWithTable();

        if ($this->shouldMountInteractsWithTable && ($defaultSortColumn = $this->table->getDefaultSortColumn()) && $this->table->getColumn($defaultSortColumn) !== null) {
            $this->tableSortColumn = $defaultSortColumn;
            $this->tableSortDirection = $this->table->getDefaultSortDirection();
        }
    }
```

This feels very hacky, though

**Reply by SanderMuller** [reply]
@danharrin I like that icon, it's nice and feels like an improvement. However, it doesn't change that you have to click the column twice before it changes the sorting

**Comment by judgej** [comment]
I can't see how that would work from the docs (v2 or v3), but I'll have a dig through the code to see how.

**Comment by atorscho** [comment]
For v2: https://filamentphp.com/docs/2.x/admin/resources/listing-records#customizing-the-eloquent-query
For v3: https://filamentphp.com/docs/3.x/panels/resources/listing-records#customizing-the-table-eloquent-query

I was able to change the default ordering thanks to this doc.

**Comment by judgej** [comment]
Customising the eloquent query is good for an underlying sorting, that the filamant table sorting sits on top of. Just be aware that the sorting column the user then selects in the UI does not *replace* the default sorting you add to the eloquent query. So if you order by post date on your eloquent query, and the user selects to order by the post title, then the result will be post titles ordered *within* post dates.

I've found this works quite well for my purposes:

```php
    public function mount()
    {
        if ($this->getTableSortColumn() === null) {
            $this->sortTable('created_at', 'desc');
        }
    }
```

This is a filament table component in a Livewire app, not an admin page.

What this does is check whether the user has arrived on the page with a default sorting column. If they have not, then set the column and order to sort by. This has the advantage that it orders the table, shows the ordering on the column being sorted (so it's not hidden from the user), and sets the URI query string, assuming [you have set them up](https://filamentphp.com/docs/2.x/admin/resources/listing-records#customizing-the-query-string).

I'm not sure how it interacts with serssion-based saving of table filters - it probably overrides those.

Doing the following instead has the same effect *except* the sorting is not shown to the user in the table UI; it's just "magically" sorted, plus it probably interacts with session-saved table filters differently:

```php
    protected function getDefaultTableSortColumn(): ?string
    {
        return 'created_at';
    }

    protected function getDefaultTableSortDirection(): ?string
    {
        return 'desc';
    }
```

**Reply by gludovatza** [reply]
In Filament v3.1.6, add to the List[ResourceName] class properties and not functions:
```
public ?string $tableSortColumn = 'created_at';

public ?string $tableSortDirection = 'desc';
```

**Reply by danharrin** [reply]
Actually, v3 has `$table->defaultSort()`

**Reply by markosole** [reply]
I've found that adding`->defaultSort('created_at', 'desc')` to `->columns([.....])->defaultSort('created_at', 'desc')`  allows manual re-sorting by any other column, which is great. While `public ?string $tableSortColumn = 'created_at';` does not allow that.
